python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1110 @@
+
+    cdef void init(self, shared_ptr[CScanner]& sp):
+        self.wrapped = sp
+        self.scanner = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CScanner]& sp):
+        cdef Scanner self = Scanner.__new__(Scanner)
+        self.init(sp)
+        return self
+
+    def scan(self):
+        """Returns a stream of ScanTasks
+
+        The caller is responsible to dispatch/schedule said tasks. Tasks should
+        be safe to run in a concurrent fashion and outlive the iterator.
+
+        Returns
+        -------
+        scan_tasks : iterator of ScanTask
+        """
+        cdef:
+            CScanTaskIterator iterator
+            shared_ptr[CScanTask] task
+
+        iterator = move(GetResultValue(move(self.scanner.Scan())))
+
+        while True:
+            iterator.Next(&task)
+            if task.get() == nullptr:
+                raise StopIteration()
+            else:
+                yield ScanTask.wrap(task)
+
+    def to_table(self):
+        """Convert a Scanner into a Table.
+
+        Use this convenience utility with care. This will serially materialize
+        the Scan result in memory before creating the Table.
+
+        Returns
+        -------
+        table : Table
+        """
+        cdef CResult[shared_ptr[CTable]] result
+
+        with nogil:
+            result = self.scanner.ToTable()
+
+        return pyarrow_wrap_table(GetResultValue(result))
+
+
+cdef class Expression:
+
+    cdef:
+        shared_ptr[CExpression] wrapped
+        CExpression* expression
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__)
+
+    cdef void init(self, const shared_ptr[CExpression]& sp):
+        self.wrapped = sp
+        self.expression = sp.get()
+
+    @staticmethod
+    cdef wrap(const shared_ptr[CExpression]& sp):
+        cdef Expression self
+
+        typ = sp.get().type()
+        if typ == CExpressionType_FIELD:
+            self = FieldExpression.__new__(FieldExpression)
+        elif typ == CExpressionType_SCALAR:
+            self = ScalarExpression.__new__(ScalarExpression)
+        elif typ == CExpressionType_NOT:
+            self = NotExpression.__new__(NotExpression)
+        elif typ == CExpressionType_CAST:
+            self = CastExpression.__new__(CastExpression)
+        elif typ == CExpressionType_AND:
+            self = AndExpression.__new__(AndExpression)
+        elif typ == CExpressionType_OR:
+            self = OrExpression.__new__(OrExpression)
+        elif typ == CExpressionType_COMPARISON:
+            self = ComparisonExpression.__new__(ComparisonExpression)
+        elif typ == CExpressionType_IS_VALID:
+            self = IsValidExpression.__new__(IsValidExpression)
+        elif typ == CExpressionType_IN:
+            self = InExpression.__new__(InExpression)
+        # TODO(kszucs): implement it
+        # elif typ == CExpressionType_CUSTOM:
+        #     self = CustomExpression.__new__(CustomExpression)
```

Comment:
For the record, in C++ `CustomExpression` is an abstract class

Reply:
I was thinking of a python side abstract class for it, but we'll handle it as a follow up. Jira needs to be created.